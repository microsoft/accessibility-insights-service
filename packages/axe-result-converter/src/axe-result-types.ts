// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import axe from 'axe-core';

export declare type SelectorType = 'xpath' | 'css';

export interface AxeCoreResults extends axe.AxeResults {
    urls: string[];
    passes: AxeResult[];
    violations: AxeResult[];
    incomplete: AxeResult[];
    inapplicable: AxeResult[];
}

export interface AxeResult extends axe.Result {
    urls: string[];
    nodes: AxeNodeResult[];
}

export interface AxeNodeResult extends axe.NodeResult {
    selectors: Selector[];
    fingerprint: string;
}

export interface Selector {
    selector: string;
    type: SelectorType;
}
