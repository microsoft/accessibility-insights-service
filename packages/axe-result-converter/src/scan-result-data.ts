// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export interface ScanResultData {
    baseUrl: string;
    basePageTitle: string;
    scanEngineName: string;
    axeCoreVersion: string;
    browserUserAgent: string;
    scanStarted: Date;
    scanEnded: Date;
}
