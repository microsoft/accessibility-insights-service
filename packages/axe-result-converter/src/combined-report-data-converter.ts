// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    CombinedReportParameters,
    AxeRuleData,
    FailureData,
    HowToFixData,
    FormattedCheckResult,
    FailuresGroup,
} from 'accessibility-insights-report';
import axe from 'axe-core';
import _ from 'lodash';
import { ScanResultData } from './scan-result-data';
import { AxeResult, AxeNodeResult, AxeCoreResults } from './axe-result-types';

interface FailureReportData {
    urlsCount: number;
    failureData: FailureData[];
}

export class CombinedReportDataConverter {
    public convert(axeResults: AxeCoreResults, scanResultData: ScanResultData): CombinedReportParameters {
        const failed = this.getFailureData(axeResults.violations);
        const failedByRule = this.groupFailureDataByRule(failed.failureData);
        const passed = this.getAxeRuleData(axeResults.passes);
        const inapplicable = this.getAxeRuleData(axeResults.inapplicable);
        const resultCounts = {
            failedUrls: failed.urlsCount,
            passedUrls: passed.length,
            unscannableUrls: inapplicable.length,
        };

        return {
            serviceName: scanResultData.scanEngineName,
            axeVersion: scanResultData.axeCoreVersion,
            userAgent: scanResultData.browserUserAgent,
            scanDetails: {
                baseUrl: scanResultData.baseUrl,
                basePageTitle: scanResultData.basePageTitle,
                scanStart: scanResultData.scanStarted,
                scanComplete: scanResultData.scanEnded,
                durationSeconds: (scanResultData.scanEnded.valueOf() - scanResultData.scanStarted.valueOf()) / 1000,
            },
            results: {
                urlResults: resultCounts,
                resultsByRule: {
                    failed: failedByRule,
                    passed: passed,
                    notApplicable: inapplicable,
                },
            },
        };
    }

    private groupFailureDataByRule(failureData: FailureData[]): FailuresGroup[] {
        const failuresGroup: FailuresGroup[] = [];
        if (failureData) {
            const failuresByRule = _.groupBy(failureData, (failure) => failure.rule.ruleId.toLowerCase());
            Object.keys(failuresByRule).map((rule) => {
                failuresGroup.push({
                    key: rule,
                    failed: failuresByRule[rule],
                });
            });
        }

        return failuresGroup.sort(this.compareFailureGroup);
    }

    private getFailureData(results: AxeResult[]): FailureReportData {
        let urlsCount = 0;
        const failureData: FailureData[] = [];
        if (results) {
            for (const result of results) {
                if (result) {
                    for (const node of result.nodes) {
                        if (node) {
                            if (result.urls) {
                                urlsCount = urlsCount + result.urls.length;
                            }

                            failureData.push({
                                urls: result.urls,
                                elementSelector: this.getElementSelector(node),
                                snippet: node.html,
                                fix: this.getNodeResult(node),
                                rule: this.getAxeRuleDataForResult(result),
                            });
                        }
                    }
                }
            }
        }

        return {
            urlsCount,
            failureData,
        };
    }

    private getAxeRuleData(results: axe.Result[]): AxeRuleData[] {
        const axeRuleData: AxeRuleData[] = [];
        if (results) {
            for (const result of results) {
                const data = this.getAxeRuleDataForResult(result);
                if (data) {
                    axeRuleData.push(data);
                }
            }
        }

        return axeRuleData;
    }

    private getNodeResult(node: axe.NodeResult): HowToFixData {
        if (node) {
            return {
                any: this.getNodeCheckResults(node.any),
                all: this.getNodeCheckResults(node.all),
                none: this.getNodeCheckResults(node.none),
                failureSummary: node.failureSummary,
            };
        }

        return undefined;
    }

    private getNodeCheckResults(nodeCheckResults: axe.CheckResult[]): FormattedCheckResult[] {
        const checkResults: FormattedCheckResult[] = [];
        if (nodeCheckResults) {
            for (const nodeCheckResult of nodeCheckResults) {
                checkResults.push({
                    id: nodeCheckResult?.id,
                    message: nodeCheckResult?.message,
                    data: nodeCheckResult?.data,
                });
            }
        }

        return checkResults;
    }

    private getAxeRuleDataForResult(result: axe.Result): AxeRuleData {
        if (result) {
            return {
                ruleId: result.id,
                tags: result.tags,
                description: result.description,
                ruleUrl: result.helpUrl,
            };
        }

        return undefined;
    }

    private getElementSelector(node: AxeNodeResult): string {
        if (node && node.selectors) {
            const css = node.selectors.find((selector) => selector.type === 'css');
            if (css) {
                return css.selector;
            } else {
                return node.selectors.find((selector) => selector.type === 'xpath')?.selector;
            }
        }

        return undefined;
    }

    private compareFailureGroup(group1: FailuresGroup, group2: FailuresGroup): number {
        if (group1.failed?.length < group2.failed?.length) {
            return 1;
        }

        if (group1.failed?.length > group2.failed?.length) {
            return -1;
        }

        return 0;
    }
}
