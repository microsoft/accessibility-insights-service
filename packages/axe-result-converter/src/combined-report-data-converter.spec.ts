// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { CombinedReportDataConverter } from './combined-report-data-converter';
import { AxeResult, AxeNodeResult, AxeCoreResults } from './axe-result-types';
import { ScanResultData } from './scan-result-data';

let combinedReportDataConverter: CombinedReportDataConverter;

const scanResultData = {
    baseUrl: 'baseUrl',
    basePageTitle: 'basePageTitle',
    scanEngineName: 'scanEngineName',
    axeCoreVersion: 'axeCoreVersion',
    browserUserAgent: 'browserUserAgent',
    scanStarted: new Date(1000),
    scanEnded: new Date(60000),
} as ScanResultData;
const accumulatedNodeFn = (ruleId: string, nodeId: string) => {
    return {
        any: [
            {
                id: `check-id-${nodeId}`,
                data: `check-data-${nodeId}`,
                message: `check-message-${nodeId}`,
            },
        ],
        all: [],
        none: [],
        failureSummary: `failureSummary-${nodeId}`,
        html: `snippet-${nodeId}`,
        target: [`selector-${nodeId}`],
        selectors: [{ selector: `selector-${nodeId}`, type: 'css' }],
        fingerprint: `${ruleId}|snippet-${nodeId}|selector-${nodeId}`,
    } as AxeNodeResult;
};
const accumulatedResultsFn = (ruleId: string, data: { urls: string[]; nodeId?: string }) =>
    [
        {
            id: `id-${ruleId}`,
            tags: [`tag-${ruleId}`],
            description: `description-${ruleId}`,
            helpUrl: `helpUrl-${ruleId}`,
            urls: data.urls,
            nodes: data.nodeId ? [accumulatedNodeFn(`id-${ruleId}`, data.nodeId)] : [],
        },
    ] as AxeResult[];

describe(CombinedReportDataConverter, () => {
    beforeEach(() => {
        combinedReportDataConverter = new CombinedReportDataConverter();
    });

    it('convert axe results to combined report data', () => {
        const violations = [
            ...accumulatedResultsFn('rule-1', { urls: ['url-11'], nodeId: 'node-11' }),
            ...accumulatedResultsFn('rule-2', { urls: ['url-21', 'url-22'], nodeId: 'node-12' }),
            ...accumulatedResultsFn('rule-2', { urls: ['url-21'], nodeId: 'node-22' }),
            ...accumulatedResultsFn('rule-3', { urls: ['url-31', 'url-32', 'url-33'], nodeId: 'node-31' }),
            ...accumulatedResultsFn('rule-3', { urls: ['url-31', 'url-32'], nodeId: 'node-32' }),
            ...accumulatedResultsFn('rule-3', { urls: ['url-31'], nodeId: 'node-33' }),
        ];
        const passes = [
            ...accumulatedResultsFn('rule-21', { urls: ['url-21'], nodeId: 'node-21' }),
            ...accumulatedResultsFn('rule-22', { urls: ['url-22'], nodeId: 'node-22' }),
        ];
        const incomplete = [
            ...accumulatedResultsFn('rule-31', { urls: ['url-31'], nodeId: 'node-31' }),
            ...accumulatedResultsFn('rule-32', { urls: ['url-32'], nodeId: 'node-32' }),
        ];
        const inapplicable = [
            ...accumulatedResultsFn('rule-41', { urls: ['url-41'] }),
            ...accumulatedResultsFn('rule-42', { urls: ['url-42'] }),
        ];
        const axeCoreResults = {
            violations,
            passes,
            incomplete,
            inapplicable,
        } as AxeCoreResults;

        const combinedReportData = combinedReportDataConverter.convert(axeCoreResults, scanResultData);

        expect(combinedReportData).toMatchSnapshot();
    });
});
