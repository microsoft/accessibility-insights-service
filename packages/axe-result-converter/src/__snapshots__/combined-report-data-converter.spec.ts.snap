// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CombinedReportDataConverter convert axe results to combined report data 1`] = `
Object {
  "axeVersion": "axeCoreVersion",
  "results": Object {
    "resultsByRule": Object {
      "failed": Array [
        Object {
          "failed": Array [
            Object {
              "elementSelector": "selector-node-31",
              "fix": Object {
                "all": Array [],
                "any": Array [
                  Object {
                    "data": "check-data-node-31",
                    "id": "check-id-node-31",
                    "message": "check-message-node-31",
                  },
                ],
                "failureSummary": "failureSummary-node-31",
                "none": Array [],
              },
              "rule": Object {
                "description": "description-rule-3",
                "ruleId": "id-rule-3",
                "ruleUrl": "helpUrl-rule-3",
                "tags": Array [
                  "tag-rule-3",
                ],
              },
              "snippet": "snippet-node-31",
              "urls": Array [
                "url-31",
                "url-32",
                "url-33",
              ],
            },
            Object {
              "elementSelector": "selector-node-32",
              "fix": Object {
                "all": Array [],
                "any": Array [
                  Object {
                    "data": "check-data-node-32",
                    "id": "check-id-node-32",
                    "message": "check-message-node-32",
                  },
                ],
                "failureSummary": "failureSummary-node-32",
                "none": Array [],
              },
              "rule": Object {
                "description": "description-rule-3",
                "ruleId": "id-rule-3",
                "ruleUrl": "helpUrl-rule-3",
                "tags": Array [
                  "tag-rule-3",
                ],
              },
              "snippet": "snippet-node-32",
              "urls": Array [
                "url-31",
                "url-32",
              ],
            },
            Object {
              "elementSelector": "selector-node-33",
              "fix": Object {
                "all": Array [],
                "any": Array [
                  Object {
                    "data": "check-data-node-33",
                    "id": "check-id-node-33",
                    "message": "check-message-node-33",
                  },
                ],
                "failureSummary": "failureSummary-node-33",
                "none": Array [],
              },
              "rule": Object {
                "description": "description-rule-3",
                "ruleId": "id-rule-3",
                "ruleUrl": "helpUrl-rule-3",
                "tags": Array [
                  "tag-rule-3",
                ],
              },
              "snippet": "snippet-node-33",
              "urls": Array [
                "url-31",
              ],
            },
          ],
          "key": "id-rule-3",
        },
        Object {
          "failed": Array [
            Object {
              "elementSelector": "selector-node-12",
              "fix": Object {
                "all": Array [],
                "any": Array [
                  Object {
                    "data": "check-data-node-12",
                    "id": "check-id-node-12",
                    "message": "check-message-node-12",
                  },
                ],
                "failureSummary": "failureSummary-node-12",
                "none": Array [],
              },
              "rule": Object {
                "description": "description-rule-2",
                "ruleId": "id-rule-2",
                "ruleUrl": "helpUrl-rule-2",
                "tags": Array [
                  "tag-rule-2",
                ],
              },
              "snippet": "snippet-node-12",
              "urls": Array [
                "url-21",
                "url-22",
              ],
            },
            Object {
              "elementSelector": "selector-node-22",
              "fix": Object {
                "all": Array [],
                "any": Array [
                  Object {
                    "data": "check-data-node-22",
                    "id": "check-id-node-22",
                    "message": "check-message-node-22",
                  },
                ],
                "failureSummary": "failureSummary-node-22",
                "none": Array [],
              },
              "rule": Object {
                "description": "description-rule-2",
                "ruleId": "id-rule-2",
                "ruleUrl": "helpUrl-rule-2",
                "tags": Array [
                  "tag-rule-2",
                ],
              },
              "snippet": "snippet-node-22",
              "urls": Array [
                "url-21",
              ],
            },
          ],
          "key": "id-rule-2",
        },
        Object {
          "failed": Array [
            Object {
              "elementSelector": "selector-node-11",
              "fix": Object {
                "all": Array [],
                "any": Array [
                  Object {
                    "data": "check-data-node-11",
                    "id": "check-id-node-11",
                    "message": "check-message-node-11",
                  },
                ],
                "failureSummary": "failureSummary-node-11",
                "none": Array [],
              },
              "rule": Object {
                "description": "description-rule-1",
                "ruleId": "id-rule-1",
                "ruleUrl": "helpUrl-rule-1",
                "tags": Array [
                  "tag-rule-1",
                ],
              },
              "snippet": "snippet-node-11",
              "urls": Array [
                "url-11",
              ],
            },
          ],
          "key": "id-rule-1",
        },
      ],
      "notApplicable": Array [
        Object {
          "description": "description-rule-41",
          "ruleId": "id-rule-41",
          "ruleUrl": "helpUrl-rule-41",
          "tags": Array [
            "tag-rule-41",
          ],
        },
        Object {
          "description": "description-rule-42",
          "ruleId": "id-rule-42",
          "ruleUrl": "helpUrl-rule-42",
          "tags": Array [
            "tag-rule-42",
          ],
        },
      ],
      "passed": Array [
        Object {
          "description": "description-rule-21",
          "ruleId": "id-rule-21",
          "ruleUrl": "helpUrl-rule-21",
          "tags": Array [
            "tag-rule-21",
          ],
        },
        Object {
          "description": "description-rule-22",
          "ruleId": "id-rule-22",
          "ruleUrl": "helpUrl-rule-22",
          "tags": Array [
            "tag-rule-22",
          ],
        },
      ],
    },
    "urlResults": Object {
      "failedUrls": 10,
      "passedUrls": 2,
      "unscannableUrls": 2,
    },
  },
  "scanDetails": Object {
    "basePageTitle": "basePageTitle",
    "baseUrl": "baseUrl",
    "durationSeconds": 59,
    "scanComplete": 1970-01-01T00:01:00.000Z,
    "scanStart": 1970-01-01T00:00:01.000Z,
  },
  "serviceName": "scanEngineName",
  "userAgent": "browserUserAgent",
}
`;
